### PR TITLE
[MM-20832] prevent dialog stacking, move to promises

### PR DESCRIPTION
**Pull request by Willyfrog**
Before submitting, please confirm you've
 - [x] executed `npm run lint:js` for proper code formatting

Please provide the following information:

**Summary**
whenever there is no connection stablished, a new connection retry is launched every few seconds, if that fails a new dialog was being presented to the user, which might end up in a stack of errors. this PR only allows one dialog per instance and error-type.

Also, on electron6 the api for the dialog moved to promises, so I updated the code for that.

**Issue link**
[MM-20832](https://mattermost.atlassian.net/browse/MM-20832)
 
**Test Cases**
change the date on your machine to a year in the future, connect to community and the dialog should show up. waiting for several minutes shouldn't create new dialogs.

### CODE REVIEW

File: src/main.js [hunk 1]
```diff
@@ -344,37 +345,56 @@ function handleAppCertificateError(event, webContents, url, error, certificate,
...
+        ({response}) => {
+          if (response === 0) {
+            dialog.showMessageBox(mainWindow, {
+              title: 'Certificate Not Trusted',
+              message: `Certificate from "${certificate.issuerName}" is not trusted.`,
+              detail,
+              type: 'error',
+              buttons: ['Trust Insecure Certificate', 'Cancel Connection'],
+              cancelId: 1,
+            }).then(({response: responseTwo}) => {
```
Comment (lieut-data) [inline src/main.js]: 0/5, but would it make sense to chain the promises together vs. nesting?

File: src/main.js [hunk 2]
```diff
@@ -344,37 +345,56 @@ function handleAppCertificateError(event, webContents, url, error, certificate,
...
+              if (responseTwo === 0) {
+                certificateStore.add(url, certificate);
+                certificateStore.save();
+                certificateErrors.get(errorID)(true);
```
Comment (lieut-data) [inline src/main.js]: Without context, it's not immediately obvious why a "certificate error" is callable. Is there a naming opportunity here?

File: src/main.js (hunk 1 above)
Comment (Willyfrog) [reply src/main.js]: changed to certificateErrorCallbacks, originally only stored if the error was being handled, but I later changed my mind into using it to store the callback

File: src/main.js (hunk 2 above)
Comment (Willyfrog) [reply src/main.js]: Definitely, changed to certificateErrorCallbacks

File: src/main.js [hunk 3]
```diff
@@ -345,56 +345,54 @@ function handleAppCertificateError(event, webContents, url, error, certificate,
...
+            title: 'Certificate Not Trusted',
+            message: `Certificate from "${certificate.issuerName}" is not trusted.`,
+            detail,
+            type: 'error',
+            buttons: ['Trust Insecure Certificate', 'Cancel Connection'],
+            cancelId: 1,
+          });
+        }
+        return {response};
       }).then(
```
Comment (lieut-data) [inline src/main.js]: nit: might want to outdent this line, to line up with line 366.

**Comment by jgilliam17** [comment]
@Willyfrog I tested using the steps from the jira issue and it works fine. Upon return to the app and reconnecting the internet red websocket banner is self dismissed.

While testing using `test cases` listed in the description here in the PR 
`change the date on your machine to a year in the future, connect to community and the dialog should show up. waiting for several minutes shouldn't create new dialogs.`
There is an immediate "Your session has expired"
I can log back in using email & password, app displays "Please check connection, Mattermost unreachable" 
Not able to log in using OneLogin, "Cannot connect to Mattermost" asking user to check connection, URL and option to open in browser.
Is this expected?

**Comment by Willyfrog** [comment]
I think it is, by moving forward into the future, any certs will probably be expired and most connections will be rejected by default as not secure. Have you tried using email+password?

**Comment by jgilliam17** [comment]
> I think it is, by moving forward into the future, any certs will probably be expired and most connections will be rejected by default as not secure. Have you tried using email+password?

Yes, I tried email+password and I can log in afterwards with app's banner displaying "Please check connection, Mattermost unreachable"
No popup dialogs though, which I guess is expected with the banner already present.

**Comment by Willyfrog** [comment]
Mannually cherry-picked

**Pull request merged**